### PR TITLE
Minor refactoring to the MavenCoordinate class

### PR DESCRIPTION
**Pull request by mir-am**
## Description
This PR makes the following changes:
- Creates a separate class for `MavenResolver` and renames it to `MavenArtifactDownloader` to avoid confusion with the Graph `MavenResolver` in core.
- Move `MavenCoordinate` and `MavenArtifactDownloader` tests to core from the OPAL module.

## Motivation and context
As stated in #338, it is better to separate the JAR downloads logic from the `MavenCoordinate` class and also moves its tests to core.

## Testing
Unit tests for `MavenCoordinate` and the JAR downloader pass successfully.

## Task list  
- [x] Separate the JAR downloader logic from `MavenCoordinate`
- [x] Move unit tests for MavenCoordinate and `MavenArtifactDownloader` to core.

### CODE REVIEW

File: analyzer/javacg-opal/src/test/java/eu/fasten/analyzer/javacgopal/data/MavenCoordinateTest.java [hunk 1]
```diff
@@ -1,115 +0,0 @@
-/*
```
Comment (proksch) [inline analyzer/javacg-opal/src/test/java/eu/fasten/analyzer/javacgopal/data/MavenCoordinateTest.java:1]: Why has this file been deleted? It is just in the wrong place... Instead of deleting it, it should be moved to FASTEN core (where the Maven Coordinate is located).

File: core/src/main/java/eu/fasten/core/data/opal/MavenArtifactDownloader.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+            "aar", "car", "nar", "kar"};
+    private static final String[] defaultPackaging = {"zip", "aar", "tar.gz", "jar"};
+
+    /**
+     * Download a JAR file indicated by the provided Maven coordinate.
+     *
+     * @param mavenCoordinate A Maven coordinate in the for "groupId:artifactId:version"
+     * @return A temporary file on the filesystem
+     */
+    public File downloadArtifact(final MavenCoordinate mavenCoordinate, String artifactRepo)
```
Comment (proksch) [inline core/src/main/java/eu/fasten/core/data/opal/MavenArtifactDownloader.java:33]: I would suggest to break this method down into several smaller helper methods... the complexity of the method goes through the roof!

File: core/src/main/java/eu/fasten/core/data/opal/MavenArtifactDownloader.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+            in.close();
+
+            return Optional.of(new File(tempFile.toAbsolutePath().toString()));
+        } catch (IOException e) {
+            if (tempFile != null) {tempFile.toFile().delete();}
+            throw new MissingArtifactException(e.getMessage(), e.getCause());
+        }
+    }
+
+    private long computeDurationInMs(long startTime) {
```
Comment (proksch) [inline core/src/main/java/eu/fasten/core/data/opal/MavenArtifactDownloader.java:117]: Taking the time for the download should be removed, this is unrelated for the task at hand. If this is necessary for a specific case, I would solve it through implementing through subclassing and overriding the corresponding method (the override can then instruments it with timing logic).

File: core/src/main/java/eu/fasten/core/data/opal/MavenArtifactDownloader.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+            long startTime = System.nanoTime();
+
+            try {
+                if (Arrays.asList(packaging).contains(mavenCoordinate.getPackaging())) {
+                    found = true;
+                    jar = httpGetFile(mavenCoordinate
+                            .toProductUrl(repos.get(i), mavenCoordinate.getPackaging()));
+                }
+            } catch (MissingArtifactException e) {
```
Comment (proksch) [inline core/src/main/java/eu/fasten/core/data/opal/MavenArtifactDownloader.java:51]: As we are cycling through multiple repositories, the log message should boil down to a single info entry that explains that artifacts was not found in repo XY and that we retry the next one... if it is not found in any repo, we can throw a `MissingArtifactException` at the end.

Related to that, I noticed that several repos are being tried before Maven Central, but as it is the most common, I would argue that Maven Central should be the first hit in the list... I am not sure though where in the code the repos are hard-coded though.

**Review by proksch — CHANGES_REQUESTED** [review]
Good first step, but more refactoring is required.

File: analyzer/javacg-opal/src/test/java/eu/fasten/analyzer/javacgopal/data/MavenCoordinateTest.java (hunk 1 above)
Comment (mir-am) [reply analyzer/javacg-opal/src/test/java/eu/fasten/analyzer/javacgopal/data/MavenCoordinateTest.java:1]: Sorry, forgot to push the tests for them in core. It's here a0ba202.

File: core/src/main/java/eu/fasten/core/data/opal/MavenArtifactDownloader.java (hunk 2 above)
Comment (mir-am) [reply core/src/main/java/eu/fasten/core/data/opal/MavenArtifactDownloader.java]: Addressed in 5f4cad7.

File: analyzer/javacg-opal/src/test/java/eu/fasten/analyzer/javacgopal/CGandStitchingTest.java [hunk 5]
```diff
@@ -340,8 +341,7 @@ private ExtendedRevisionJavaCallGraph getRCG(String coordStr)
         throws OPALException, MissingArtifactException {
         final var coord = MavenCoordinate.fromString(coordStr,"jar");
 
-        final var dep1 =
-            new MavenCoordinate.MavenResolver().downloadArtifact(coord,"jar");
+        final var dep1 = new MavenArtifactDownloader().downloadArtifact(coord,"jar");
```
[the remark this answers is not in the document]
Comment (mir-am) [reply analyzer/javacg-opal/src/test/java/eu/fasten/analyzer/javacgopal/CGandStitchingTest.java]: Done!

File: core/src/main/java/eu/fasten/core/data/opal/MavenArtifactDownloader.java (hunk 3 above)
Comment (mir-am) [reply core/src/main/java/eu/fasten/core/data/opal/MavenArtifactDownloader.java:151]: I am not sure whether we can remove the timing info there. Wouter changed the format of OPAL's logs and also, IIRC included timing for his monitoring solution.

File: core/src/main/java/eu/fasten/core/data/opal/MavenArtifactDownloader.java (hunk 4 above)
Comment (mir-am) [reply core/src/main/java/eu/fasten/core/data/opal/MavenArtifactDownloader.java:80]: Source repos are specified in the input records of OPAL, i.e., the `artifactRepository` field. I think it tries the Maven Central repo first most of the time as most input records have Maven Central defined as the source repo.

**Comment by mir-am** [comment]
@proksch, thanks for your contribution.
Deleted tests in a8dd5e4 are NOT duplicate. I moved them from the OPAL module to the core module.
